### PR TITLE
SRV-309 Process campaign targeting

### PR DESCRIPTION
**Pull request by PawelPodkalicki**

### CODE REVIEW

File: app/Http/Controllers/Manager/CampaignsController.php [hunk 1]
```diff
@@ -237,8 +253,15 @@ public function edit(Request $request, int $campaignId): JsonResponse
...
-        $input['targeting_excludes'] = $request->input('campaign.targeting.excludes');
+
+        $targetingProcessor = new TargetingProcessor($this->configurationRepository->fetchTargetingOptions());
+
+        $input['targeting_requires'] = $targetingProcessor->processTargeting(
+            $request->input('campaign.targeting.requires')
+        );
+        $input['targeting_excludes'] = $targetingProcessor->processTargeting(
+            $request->input('campaign.targeting.excludes')
+        );
```
Comment (c3zi) [inline app/Http/Controllers/Manager/CampaignsController.php]: The same code block is used twice. I'd suggest to move it to a dedicated method (``` $input = $this->processTargeting($input) ```).

File: app/Http/Requests/Campaign/TargetingProcessor.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                if (count($processed) > 0) {
+                    $groupsProcessed[$key] = $processed;
+                }
+
+                break;
+            }
+        }
+
+        return $groupsProcessed;
+    }
```
Comment (c3zi) [inline app/Http/Requests/Campaign/TargetingProcessor.php:66]: Could be changed from O(n^2) to O(n) by preparing a temporary array :)

File: tests/app/Http/Requests/Campaign/TargetingProcessorTest.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        ],
+        "browser": [
+            "opera"
+        ]
+    }
+}
+JSON
+            ,
+            true
+        );
```
Comment (c3zi) [inline tests/app/Http/Requests/Campaign/TargetingProcessorTest.php]: Maybe just `array`? I understand intentions but array would look "cleaner".

**Review by c3zi — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Manager/CampaignsController.php (hunk 1 above)
Comment (PawelPodkalicki) [reply app/Http/Controllers/Manager/CampaignsController.php]: sure

File: tests/app/Http/Requests/Campaign/TargetingProcessorTest.php (hunk 3 above)
Comment (PawelPodkalicki) [reply tests/app/Http/Requests/Campaign/TargetingProcessorTest.php]: I will change

**Review by c3zi — DISMISSED** [review]

**Review by c3zi — APPROVED** [review]

**Pull request merged**